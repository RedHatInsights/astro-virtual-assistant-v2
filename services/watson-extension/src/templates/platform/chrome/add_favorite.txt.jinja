{%- if already -%}
Service [{{service}}]({{href}}) is already favorited.
{%- else -%}
The [{{service}}]({{href}}) ({{group}}) service is now in your Favorites!
{%- endif -%}
