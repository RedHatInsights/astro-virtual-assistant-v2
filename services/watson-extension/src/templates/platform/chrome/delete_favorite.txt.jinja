{%- if not already -%}
The {{service}} wasn't in your Favorites.
{%- else -%}
The [{{service}}]({{href}}) ({{group}}) service is no longer favorited.
{%- endif -%}
