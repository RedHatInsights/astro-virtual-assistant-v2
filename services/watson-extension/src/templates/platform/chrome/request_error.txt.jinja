I'm sorry. I couldn't remove the [{{service}}]({{href}}) ({{group}}) service from your Favorites.
