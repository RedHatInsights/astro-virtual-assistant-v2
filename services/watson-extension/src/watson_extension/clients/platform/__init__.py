from dataclasses import dataclass


@dataclass
class IntegrationInfo:
    name: str
    enabled: bool
    type: str
    group: str
    id: str
