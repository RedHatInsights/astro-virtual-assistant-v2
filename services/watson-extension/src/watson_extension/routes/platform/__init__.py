from quart import Blueprint
from . import chrome

blueprint = Blueprint("platform", __name__, url_prefix="/platform")

blueprint.register_blueprint(chrome.blueprint)
