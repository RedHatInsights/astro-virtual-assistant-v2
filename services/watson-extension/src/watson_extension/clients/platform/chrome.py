import abc
import json
import injector
from dataclasses import dataclass
from typing import List, Optional

from common.identity import AbstractUserIdentityProvider
from common.platform_request import AbstractPlatformRequest
from watson_extension.clients import ChromeServiceURL


@dataclass
class Favorite:
    id: str
    pathname: str
    favorite: bool
    user_identity_id: str


@dataclass
class User:
    account_id: str
    first_login: bool
    day_one: bool
    last_login: str
    last_visited_pages: List[dict]
    favorite_pages: List[Favorite]
    visited_bundles: dict


@dataclass
class Link:
    id: Optional[str]
    title: str
    alt_title: Optional[List[str]]
    links: Optional["Link"]
    app_id: Optional[str] = None
    is_group: bool = False
    is_external: bool = False
    href: Optional[str] = None


@dataclass
class Service:
    description: str
    id: str
    links: Optional[List[dict]]
    title: str
    href: Optional[str]


class ChromeServiceClient(abc.ABC):
    @abc.abstractmethod
    async def get_user(self) -> User: ...

    @abc.abstractmethod
    async def get_generated_services(self) -> List[Service]: ...

    @abc.abstractmethod
    async def modify_favorite_service(
        self, service, favorite=True
    ) -> List[Favorite]: ...


class ChromeServiceClientHttp(ChromeServiceClient):
    def __init__(
        self,
        chrome_url: injector.Inject[ChromeServiceURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.chrome_url = chrome_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def get_user(self):
        # struggles to be a json response, manually doing it here
        response = await self.platform_request.get(
            self.chrome_url,
            "/api/chrome-service/v1/user",
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        data = json.loads(await response.text())["data"]
        return User(
            account_id=data["accountId"],
            first_login=data["firstLogin"],
            day_one=data["dayOne"],
            last_login=data["lastLogin"],
            last_visited_pages=data["lastVisitedPages"],
            favorite_pages=[
                Favorite(
                    id=fp["id"],
                    pathname=fp["pathname"],
                    favorite=fp["favorite"],
                    user_identity_id=fp["userIdentityId"],
                )
                for fp in data.get("favoritePages", [])
            ],
            visited_bundles=data.get("visitedBundles", {}),
        )

    async def get_generated_services(self) -> List[Service]:
        response = await self.platform_request.get(
            self.chrome_url,
            "/api/chrome-service/v1/static/stable/prod/services/services-generated.json",
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        data = await response.json()
        return [
            Service(
                description=service.get("description", ""),
                id=service.get("id", ""),
                links=parse_links_into_obj(service.get("links", [])),
                title=service.get("title", ""),
                href=service.get("href", ""),
            )
            for service in data
        ]

    async def modify_favorite_service(self, href, favorite=True):
        response = await self.platform_request.post(
            self.chrome_url,
            "/api/chrome-service/v1/favorite-pages",
            user_identity=await self.user_identity_provider.get_user_identity(),
            json={
                "favorite": favorite,
                "pathname": href,
            },
        )
        response.raise_for_status()
        return await response.text()


def parse_links_into_obj(links) -> List[Link]:
    if not links:
        return []
    parsed_links = []
    for link in links:
        parsed_links.append(
            Link(
                id=link.get("id"),
                title=link.get("title", ""),
                alt_title=link.get("altTitle", []),
                links=parse_links_into_obj(link.get("links", [])),
                app_id=link.get("appId"),
                is_group=link.get("isGroup", False),
                is_external=link.get("isExternal", False),
                href=link.get("href", ""),
            )
        )
    return parsed_links
