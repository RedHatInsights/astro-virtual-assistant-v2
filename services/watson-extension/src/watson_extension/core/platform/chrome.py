from typing import List
import injector

from watson_extension.clients.platform.chrome import ChromeServiceClient


class ChromeServiceCore:
    def __init__(self, chrome_service_client: injector.Inject[ChromeServiceClient]):
        self.chrome_service_client = chrome_service_client
    
    async def get_favorite_options(self, favoriting=True):
        user = await self.chrome_service_client.get_user()
        services = await self.chrome_service_client.get_generated_services()
        
        #if a user is requesting to remove a service, return info on the services they have favorited
        # if they are requesting to add one, return the opposite
        if not user.favorite_pages:
            return self._get_all_service_info(services)
        
        return (
            self._get_non_favorited_services_info(self, services, user.favorite_pages) if favoriting
            else self._get_favorited_services_info(self, services, user.favorite_pages)
        )
    
    def _get_all_service_info(self, services) -> List[dict]:
        return [{
                "title": service.title,
                "href": service.href,
            } for service in services]
    
    def _get_non_favorited_services_info(self, services, favorite_pages) -> List[dict]:
        return [{
            "title": service.title,
            "href": service.href,
        } for service in services if service.href not in {fav.pathname for fav in favorite_pages}]

    def _get_favorited_services_info(self, services, favorite_pages) -> List[dict]:
        return  [{
            "title": self._get_service_by_title(fav.pathname),
            "href": fav.pathname,
        } for fav in favorite_pages]

    async def get_service_options(self):
        services = await self.chrome_service_client.get_generated_services()
        parsed = parse_generated_services(services)
        options = []
        for s in parsed:
            if s.get("href", ""):
                options.append(convert_service_to_option(s))

        return options

    async def get_user(self):
        return await self.chrome_service_client.get_user()

    async def _get_service_by_title(self, title):
        def is_synonym(service):
            return title.lower() in (t.lower() for t in service["synonyms"])

        options = await self.get_service_options()
        service = next((s for s in options if is_synonym(s)), None)

        return service["data"] if service else None

    async def _is_favorited(self, href):
        user = await self.get_user()
        if not user.favorite_pages:
            return False
        favorites = user.favorite_pages
        return any(favorite.pathname == href for favorite in favorites)

    async def get_service_data(self, title):
        service = await self._get_service_by_title(title)
        if not service:
            return None

        href = service["href"]
        favorited_already = await self._is_favorited(href)

        return {
            "service": service["title"],
            "href": href,
            "group": service["group"],
            "already": favorited_already,
        }

    async def modify_favorite_service(self, href, favorite):
        return await self.chrome_service_client.modify_favorite_service(
            href=href, favorite=favorite
        )


# Helpers
def parse_generated_services(content):
    services = []
    for category in content:
        for link in category.links:
            group_title = link.title
            if link.is_group:
                for sublink in link.links:
                    if sublink.is_external:
                        # not really a service
                        continue
                    service = {"group": group_title}

                    if sublink.href:
                        service["href"] = sublink.href
                    if sublink.title:
                        service["title"] = sublink.title
                    if sublink.app_id:
                        service["app_id"] = sublink.app_id
                    service["alt_title"] = sublink.alt_title
                    services.append(service)
    return services


def convert_service_to_option(service):
    value = {"group": service.get("group", "")}
    synonyms = [service.get("href", "")]
    value["href"] = service.get("href", "")

    if "title" in service:
        value["title"] = service["title"]
        synonyms.append(service["title"])
    if "appId" in service:
        value["app_id"] = service["appId"]
        synonyms.append(service["appId"])
    if "alt_title" in service:
        synonyms += service["alt_title"]
    return {
        "data": value,
        "synonyms": synonyms,
    }
