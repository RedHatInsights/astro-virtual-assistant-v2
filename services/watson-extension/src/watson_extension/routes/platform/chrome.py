from typing import List
import injector
from pydantic import BaseModel

from quart import Blueprint, render_template
from quart_schema import validate_response, validate_querystring, document_headers

from watson_extension.core.platform.chrome import (
    ChromeServiceCore,
)
from watson_extension.routes import RHSessionIdHeader

blueprint = Blueprint("chrome", __name__, url_prefix="/chrome")


class FavoritesRequestQuery(BaseModel):
    title: str
    favoriting: bool  # true is adding a favorite


class FavoritesResponse(BaseModel):
    response: str


class ServiceResponseItem(BaseModel):
    data: dict
    synonyms: List[str]


class ServicesResponse(BaseModel):
    response: List[ServiceResponseItem]


@blueprint.get("/services")
@validate_response(ServicesResponse)
@document_headers(RHSessionIdHeader)
async def services(
    chrome_service: injector.Inject[ChromeServiceCore],
) -> ServicesResponse:
    options = await chrome_service.get_service_options()
    return ServicesResponse(response=options)


@blueprint.post("/favorites")
@validate_querystring(FavoritesRequestQuery)
@validate_response(FavoritesResponse)
@document_headers(RHSessionIdHeader)
async def favorites(
    query_args: FavoritesRequestQuery,
    chrome_service: injector.Inject[ChromeServiceCore],
) -> FavoritesResponse:
    service_data = await chrome_service.get_service_data(
        title=query_args.title,
    )
    if not service_data:
        return FavoritesResponse(
            response=await render_template(
                "platform/chrome/service_not_found.txt.jinja",
                title=query_args.title,
            )
        )
    if service_data["already"] != query_args.favoriting:
        try:
            await chrome_service.modify_favorite_service(
                service_data["href"], favorite=query_args.favoriting
            )
        except Exception:
            return FavoritesResponse(
                response=await render_template(
                    "platform/chrome/request_error.txt.jinja",
                    **service_data,
                )
            )

    if query_args.favoriting:
        return FavoritesResponse(
            response=await render_template(
                "platform/chrome/add_favorite.txt.jinja",
                **service_data,
            )
        )
    else:
        return FavoritesResponse(
            response=await render_template(
                "platform/chrome/delete_favorite.txt.jinja",
                **service_data,
            )
        )
