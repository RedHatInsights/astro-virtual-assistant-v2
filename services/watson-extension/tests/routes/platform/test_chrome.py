from unittest.mock import MagicMock

import injector
import pytest
from quart.typing import TestClientProtocol

from watson_extension.clients.platform.chrome import (
    ChromeServiceClient,
    Favorite,
    User,
)
from ..common import app_with_blueprint

from watson_extension.routes.platform.chrome import blueprint
from ... import async_value, get_test_template


@pytest.fixture
async def chrome_client() -> MagicMock:
    chrome_client = MagicMock(ChromeServiceClient)
    chrome_client.modify_favorite_service = MagicMock(
        return_value=async_value([
            Favorite(id="asdfg", pathname="/insights/foo", favorite=True, user_identity_id="1234"),
            Favorite(id="asdfg", pathname="/insights/bar", favorite=False, user_identity_id="1234")
        ])
    )

    chrome_client.get_user = MagicMock(
        return_value=async_value(
            User(
                account_id="1234",
                first_login=False,
                day_one=False,
                last_login="2023-01-01",
                last_visited_pages=[],
                favorite_pages=[
                    Favorite(id="bar2", pathname="/insights/bar2", favorite=True, user_identity_id="1234")
                ],
                visited_bundles={},
            )
        )
    )

    chrome_client.get_generated_services = MagicMock(
        return_value=async_value([{
            "description": "foo",
            "id": "str",
            "links": [{
                "title": "barGroup",
                "isGroup": True, 
                "href": "/insights/bar",
                "group": "insights",
                "id": "barGroup",
                "links": [
                    {
                        "appId": "bar1",
                        "description": "barring this test from passing",
                        "filterable": False,
                        "href": "/insights/bar1/run-damnit",
                        "icon": "AITechnologyIcon",
                        "id": "bar1",
                        "title": "bar1"
                    },
                    {
                        "appId": "bar2",
                        "description": "barring this test from passing",
                        "filterable": False,
                        "href": "/insights/bar2/run-damnit",
                        "icon": "AITechnologyIcon",
                        "id": "bar2",
                        "isExternal": True,
                        "title": "bar2"
                    },
                    {
                        "appId": "bar3",
                        "description": "barring this test from passing",
                        "filterable": False,
                        "href": "/insights/bar2/run-damnit",
                        "icon": "AITechnologyIcon",
                        "id": "bar3",
                        "title": "bar3"
                    },
                ]
            }],
            "title": "foo",
            "href": "/insights/foo",
            "group": "insights",
        }])
    )
    return chrome_client


@pytest.fixture
async def test_client(chrome_client) -> TestClientProtocol:
    def injector_binder(binder: injector.Binder):
        binder.bind(ChromeServiceClient, chrome_client)

    return app_with_blueprint(blueprint, injector_binder).test_client()


async def test_favoriting_add_favorite(test_client, chrome_client) -> None:
    response = await test_client.post(
        "/chrome/favorites", query_string={"favoriting": True, "title": "bar1"}
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == get_test_template("platform/chrome/add_favorite.txt")


async def test_favoriting_external_service(test_client, chrome_client) -> None:
    response = await test_client.post(
        "/chrome/favorites", query_string={"favoriting": True, "title": "bar2"}
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == get_test_template("platform/chrome/external_service.txt")


async def test_favoriting_already_favorited(test_client, chrome_client) -> None:
    response = await test_client.post(
        "/chrome/favorites", query_string={"favoriting": True, "title": "bar3"}
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == get_test_template("platform/chrome/already_favorited.txt")


async def test_favoriting_not_found(test_client, chrome_client) -> None:
    response = await test_client.post(
        "/chrome/favorites", query_string={"favoriting": True, "title": "not-found"}
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == get_test_template("platform/chrome/not_found_favorite.txt")
