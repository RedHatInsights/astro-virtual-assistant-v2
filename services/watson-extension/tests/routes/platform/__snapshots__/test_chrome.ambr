# serializer version: 1
# name: test_favoriting_add_favorite
  'The [bar1](/insights/bar1/run-damnit) (barGroup) service is now in your Favorites!'
# ---
# name: test_favoriting_already_favorited
  'Service [bar3](/insights/bar3/run-damnit) is already favorited.'
# ---
# name: test_favoriting_external_service
  "Service bar2 wasn't found."
# ---
# name: test_favoriting_not_found
  "Service not-found wasn't found."
# ---
# name: test_favoriting_remove_favorite
  'The [bar3](/insights/bar3/run-damnit) (barGroup) service is no longer favorited.'
# ---
# name: test_favoriting_remove_service_not_favorited
  "The bar1 wasn't in your Favorites."
# ---
